swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -1292,7 +1403,111 @@ string ppReprType(DataFlowType t) { none() }
  * a node of type `t1` to a node of type `t2`.
  */
 pragma[inline]
-predicate compatibleTypes(DataFlowType t1, DataFlowType t2) { any() }
+predicate compatibleTypes(DataFlowType t1, DataFlowType t2) {
+  subType(t1) = subType(t2)
+  or
+  exists(BoundGenericType bound1, BoundGenericType bound2 |
+    stripType(t1.asType()).(BoundGenericType) = bound1 and
+    stripType(t2.asType()).(BoundGenericType) = bound2 and
+    bound1.getDeclaration() = bound2.getDeclaration() and
+    forall(int index | index in [0 .. bound1.getNumberOfArgTypes() - 1] |
+      isAnyOrUnboundType(bound1.getArgType(index).getCanonicalType()) or
+      isAnyOrUnboundType(bound2.getArgType(index).getCanonicalType())
+    )
+  )
```

Comment:
So just to make sure I understand this: If we have to instantiations of two generic types, then they are compatible if:
1. they are coming from the "same generic type" (i.e., the same declaration in the source file)
2. for each type argument index, one of the type arguments at this position is an any type or an unbound type

Is that correct? What happens if none of the argument types is an any type or an unbound type? Are the types then compatible by the `subType(t1) = subType(t2)` then?

Reply:
Yes. There's a potential issue if we need apply both kinds of rules to different argument type positions, though.